sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcUtils.scala

```diff
@@ -142,13 +142,17 @@ object OrcUtils extends Logging {
       reader: Reader,
       conf: Configuration): Option[(Array[Int], Boolean)] = {
     val orcFieldNames = reader.getSchema.getFieldNames.asScala
+    val forcePositionalEvolution = OrcConf.FORCE_POSITIONAL_EVOLUTION.getBoolean(conf)
```

Comment:
this doesn't look like a reasonable config for a data source, but more like a config for the upper level that maps metastore schema with the ORC files.

Do we have a good story for the schema evolution of spark file sources?

Reply:
Sorry, I'm not sure I get your point.
In `requestedColumnIds()` we map `requiredSchema` to the schema in the ORC files (`orcFieldNames`) and actually Spark is prepared that the schema in HMS and in the file doesn't always match: https://github.com/apache/spark/pull/29737/files#diff-3fb8426b690ab771c4f67f9cad336498L149 (the file is written by an old version of Hive).
It turned out that a schema mismatch can happen with newer version of Hive (columns in the file doesn't start with `_col`) too. Because simply setting `orc.force.positional.evolution=true` and then doing a column rename in Hive also results mismatch in Spark and in that case Spark returns `null`s now.
It seemed a good idea to add support for this setting to our data source but if that is not the good way to deal with the issue please let me know.

Reply:
I mean, this doesn't look like an ORC-specific thing. Does hive only do this for ORC tables but not Parquet tables?

Reply:
I see your point now, but this is an ORC specific setting in Hive. With parquet tables you get `NULL` on a renamed column:
```
> set orc.force.positional.evolution;
+--------------------------------------+
|                 set                  |
+--------------------------------------+
| orc.force.positional.evolution=true  |
+--------------------------------------+
> create external table t2 (c1 string, c2 string) stored as parquet;
> insert into t2 values ('foo', 'bar');
> alter table t2 change c1 c3 string;
> select * from t2;
+--------+--------+
| t2.c3  | t2.c2  |
+--------+--------+
| NULL   | bar    |
+--------+--------+
```